pkg/cloudprovider/huaweicloud/huaweicloud.go

```diff
@@ -940,114 +814,170 @@ func (h *CloudProvider) listenerDeploy() error {
+				HolderIdentity:       *ccmHolderIdentity,
+				LeaseDurationSeconds: 60,
+				RenewTime:            metav1.NewTime(time.Now().Add(50 * time.Second)),
+			}
+			err = leaderLock.configMapLock.Update(context.TODO(), *newLeader)
+			if err != nil {
+				klog.Errorf("failed to update leader:%v", err)
+			}
+			listener.stopListenerSlice()
+			listener.goroutinePool.Stop()
+		}
 	}, func() {
 		listener.stopListenerSlice()
 		listener.goroutinePool.Stop()
 	})
+
 	return nil
 }
 
-func leaderElection(id string, restConfig *rest.Config, recorder record.EventRecorder, onSuccess func(context.Context), onStop func()) {
-	leaseName := "endpoint-slice-listener"
-	leaseDuration := 60 * time.Second
-	renewDeadline := 50 * time.Second
-	retryPeriod := 30 * time.Second
+func buildConfigMapLock(restConfig *rest.Config, recorder record.EventRecorder) *LeaderLock {
+	id, err := os.Hostname()
+	if err != nil {
+		klog.Errorf("failed to elect leader in listener EndpointSlice : %s", err)
+		return nil
+	}
+
+	leaderLock := &LeaderLock{
+		id:        id,
+		leaseName: endpointListenerLeaseName,
 
+		leaseDuration: 60 * time.Second,
+		renewDeadline: 50 * time.Second,
+		retryPeriod:   30 * time.Second,
+
+		restConfig: restConfig,
+		recorder:   recorder,
+	}
+
+	cfgIdentity := fmt.Sprintf("%s_%s", id, string(uuid.NewUUID()))
 	configmapLock, err := resourcelock.NewFromKubeconfig(resourcelock.ConfigMapsLeasesResourceLock,
 		"kube-system",
-		leaseName,
+		leaderLock.leaseName,
 		resourcelock.ResourceLockConfig{
-			Identity:      fmt.Sprintf("%s_%s", id, string(uuid.NewUUID())),
+			Identity:      cfgIdentity,
 			EventRecorder: recorder,
 		},
 		restConfig,
-		renewDeadline)
+		leaderLock.renewDeadline)
 	if err != nil {
 		klog.Fatalf("error creating lock: %v", err)
 	}
 
+	leaderLock.configMapLock = configmapLock
+
+	return leaderLock
+}
+
+func leaderElection(leaderLock *LeaderLock, onSuccess func(context.Context), onStop func()) {
 	leaderelection.RunOrDie(context.TODO(), leaderelection.LeaderElectionConfig{
-		Lock:          configmapLock,
-		LeaseDuration: leaseDuration,
-		RenewDeadline: renewDeadline,
-		RetryPeriod:   retryPeriod,
+		Lock:          leaderLock.configMapLock,
+		LeaseDuration: leaderLock.leaseDuration,
+		RenewDeadline: leaderLock.renewDeadline,
+		RetryPeriod:   leaderLock.retryPeriod,
 		Callbacks: leaderelection.LeaderCallbacks{
 			OnStartedLeading: func(ctx context.Context) {
-				klog.V(6).Infof("[Listener EndpointSlices] leader election got: %s", id)
+				klog.V(6).Infof("[Listener EndpointSlices] leader election got: %s", leaderLock.id)
 				onSuccess(ctx)
 			},
 			OnStoppedLeading: func() {
-				klog.Infof("[Listener EndpointSlices] leader election lost: %s", id)
+				klog.Infof("[Listener EndpointSlices] leader election lost: %s", leaderLock.id)
 				onStop()
 			},
 		},
-		Name: leaseName,
+		Name: leaderLock.leaseName,
 	})
 }
+
+func checkHostName(ccmHolderIdentity, epHolderIdentity string) bool {
```

Comment:
What does `checkHostName` do?

Reply:
only when one node holds `cloud-controller-manager` and then it will have access to start `endpoint-slice-listener`